Display a notice if a field isn't available anymore

kienstra opened the pull request:
Update: This description is mainly outdated. Please see the comments below.
* This PR is the beginning of the conversation. The copy can change 😄 
* Right now, this says that maybe the pro license expired.
* But there should probably be a check that this is a pro
field.
<img width="1018" alt="new-notice" src="https://user-images.githubusercontent.com/4063887/57976342-28a45800-79a3-11e9-9c56-0ac4095590b7.png">

Right now, the link goes to:
<img width="1335" alt="link-goes-to" src="https://user-images.githubusercontent.com/4063887/57976353-4d98cb00-79a3-11e9-98fb-42d3476bdf3c.png">

Fixes #242

kienstra committed:
If the field isn't available anymore, add a notice

Right now, this says that maybe the pro license expired.
But there should probably be a check that this is a pro
field.

kienstra committed:
Change the notice to be a warning (yellow) instead of blue

This is probably more appropriate,
as it's probably something the user should fix.

kienstra commented:
This still allows editing a Pro field that's not available, but the `<select>` will default to a "Text" type:
![prof-field](https://user-images.githubusercontent.com/4063887/57976783-63f85400-79ae-11e9-8d52-a4ebc9957219.gif)


Maybe this is confusing. But it allows the user to change the field. For example, maybe they would change a Post field to a Text field with the post ID.

kienstra commented:
**Question About Copy And Link**

Hi @lukecarbis and @RobStino,
What do you think about the copy and the link location here?

>This post field type is not available. Maybe your pro license expired.

<img width="502" alt="copy-and-link" src="https://user-images.githubusercontent.com/4063887/57976796-a1f57800-79ae-11e9-946f-708609052012.png">

The link goes to the "Go Pro" page:
<img width="1335" alt="link-goes-to" src="https://user-images.githubusercontent.com/4063887/57976353-4d98cb00-79a3-11e9-98fb-42d3476bdf3c.png">

Thanks!

lukecarbis committed:
Merge branch 'develop' into fix/expired-pro-license

lukecarbis reviewed (changes requested):
@kienstra So glad you're onto this! Great PR.

Here's my suggestion on how the notice should look (using the warning dashicon). I'm not really happy with forcing a notice in that small space:
<img width="1205" alt="Screen Shot 2019-05-20 at 8 44 37 am" src="https://user-images.githubusercontent.com/1097667/57989275-cdc24c00-7adb-11e9-8e4d-064823ee9436.png">

(@RobStino feel free to chime in on this)

Also note my copy change suggestion. This is a bit shorter and to the point:
> This {field type} field requires an active pro license.

I really like linking to the Pro page. Good call.

The other nice-to-have with this PR would be to _keep_ the settings for the disabled control, but have them disabled. This is so that the values won't be lost if you save the Block, and then re-activate your pro license. (Sure, it would be trivial to re-enable the fields using the Inspector and save changes to them that way, but the point is that the field won't show in the editor or the frontend)

RobStino commented:
UI and copy is great. 
Some thoughts. On the move so sorry about the brevity. 😄
With an inactive license, what _should_ we disable. Access to the Pro field in all instances, or just the ability to edit the field in the block editor?
What if a block built with pro sits on a site for many years. Is the site owner required to have an active pro license ongoing? Even if they never build a pro block again? 
I guess it comes down to what pro is. The ability to “build blocks with pro fields” or the ability to “build and use blocks with pro fields “
I don’t know if I have a strong opinion, so super interested in your thoughts.

kienstra commented:
Hi @RobStino,
Thanks, good questions.

>What if a block built with pro sits on a site for many years. Is the site owner required to have an active pro license ongoing? Even if they never build a pro block again?

Good point. The front-end output of Pro fields will be different when the license expires. I hadn't thought of that before.

This is because the pro fields usually use an `array()` for their value. For example, the Post control saves the value as an array, and [has a validate() method](https://github.com/getblocklab/block-lab/blob/eda671f134819f5dd87af271c4d6822c7364e160/php/blocks/controls/class-post.php#L130) that parses the `array()` for `block_value()` and `block_field()`. 

But when the pro version license expires, that `validate()` method doesn't run, and the output is unexpected:

<img width="817" alt="block-field-array" src="https://user-images.githubusercontent.com/4063887/57989914-096efe80-7a67-11e9-8454-ac1e6867a7bc.png">

kienstra commented:
@lukecarbis, good points. I'm working on applying them.

kienstra committed:
Change the copy for the expired pro license

Apply Luke's suggestion for the copy,
as it's more compact.

kienstra commented:
https://github.com/getblocklab/block-lab/pull/306/commits/2ca1e2571caa0e6a8042afc7ac5a4d02bc10f44b applies Luke's suggestion for the copy:

<img width="959" alt="new-copy" src="https://user-images.githubusercontent.com/4063887/57990235-51dbeb80-7a6a-11e9-8004-e0fb00c12197.png">

lukecarbis commented:
It would be great to apply a little styling so that the icon sits left of the text, as per the screenshot I posted earlier.

kienstra commented:
>The other nice-to-have with this PR would be to keep the settings for the disabled control, but have them disabled. This is so that the values won't be lost if you save the Block, and then re-activate your pro license. (Sure, it would be trivial to re-enable the fields using the Inspector and save changes to them that way, but the point is that the field won't show in the editor or the frontend)

This is a good idea. I'm working on it now, though I haven't figured out a simple way to do this.

kienstra committed:
If the pro license expires, don't overwrite previous values

As Luke mentioned, this allows someone to
renew the pro license without losing the values.
This will disable and make readonly the fields.
readonly is useful because it allows the name
to be part of the POST request to update the block.
A disabled <input> isn't part of the POST request.

kienstra committed:
Apply a margin to separate the dashicon and the message

Apply the styling that Luke showed
in his screenshot.

kienstra commented:
https://github.com/getblocklab/block-lab/pull/306/commits/e3d26a6a6d379c1f2335731f32a548236ba878af adds a margin to separate the dashicon and the message:

<img width="910" alt="margin-here" src="https://user-images.githubusercontent.com/4063887/58002708-b6b43780-7aa4-11e9-9ba0-c4547c13569d.png">

kienstra commented:
>The other nice-to-have with this PR would be to keep the settings for the disabled control, but have them disabled. This is so that the values won't be lost if you save the Block, and then re-activate your pro license. (Sure, it would be trivial to re-enable the fields using the Inspector and save changes to them that way, but the point is that the field won't show in the editor or the frontend)

Hi @lukecarbis,
Good idea. The commits above make it possible to get the same values for pro fields when you reactivate an expired pro license:

Screenshot 1) For example, here's a Taxonomy field with the pro license activated:
<img width="1134" alt="taxonomy-field" src="https://user-images.githubusercontent.com/4063887/58003629-c6348000-7aa6-11e9-8ef1-fa22be03e12c.png">

Screenshot 2) When the pro license expires, here's what it'll look like:
<img width="1123" alt="taxonmy-here" src="https://user-images.githubusercontent.com/4063887/58003828-44912200-7aa7-11e9-97ec-b5c6cba1e806.png">

And if the pro license is reactivated, it will looks like Screenshot 1 again.

Is this what you had in mind?

kienstra committed:
Address failed Travis build by adding a @param tag

There should have been a @param tag for $control_name,
so add that.

kienstra commented on php/post-types/class-block-post.php:
```diff
@@ -578,7 +620,15 @@ public function render_fields_meta_box_row( $field, $uid = false ) {
+							<?php if ( $is_field_disabled ) : ?>
+								<input
+									name="block-is-disabled-pro-field[<?php echo esc_attr( $uid ); ?>]"
+									type="hidden"
```
Having a hidden `<input>`to show that the pro field (control) is disabled isn't ideal.
 
But because I set the `<select>` for the control type as `disabled`, its value won't be present in the POST request. So I couldn't figure out how to find if it's a pro control, without adding this hidden `<input>`.

lukecarbis replied:
Have you tested this out with checkboxes?

kienstra commented on php/post-types/class-block-post.php:
```diff
@@ -598,7 +648,8 @@ class="regular-text"
 								value="<?php echo esc_attr( $field->name ); ?>"
-								data-sync="block-fields-name-code_<?php echo esc_attr( $uid ); ?>" />
+								data-sync="block-fields-name-code_<?php echo esc_attr( $uid ); ?>"
+								<?php echo $is_field_disabled ? 'readonly="readonly"' : ''; ?> />
```
You're probably aware of this, but using `readonly` instead of `disabled` means that it will be in the `POST` request. Having the name of the control in the request is important. Otherwise, it's not saved.

lukecarbis replied:
Makes sense. Maybe we could add some CSS along the lines of:
```
.block-fields-edit input[readonly="readonly"],
.block-fields-edit textarea[readonly="readonly"] {
    color: #999;
}
```

RobStino commented:
Hey,
Testing this now. Looking good, but I did notice something odd.
Firstly, I did see that you can still add and interact as normal with a Pro field within a block when the Pro license has been deactivated. This is cool! It looks like this was already thought through and my comments further up weren't needed. :D
The odd thing I noticed was with the User field.
While the Pro license was active, block_field returned a string of the Username. So the output in the editor and on the frontend was just `Rob`.
With the license deactived it output `1,Rob` in both the editor and frontend. I'm guessing this is an array of the User ID and User Display Name.

*License Active*
![image](https://user-images.githubusercontent.com/1578148/58008917-b8c3d800-7b30-11e9-9344-8ee6b3e79d3d.png)

*License Inactive*
![image](https://user-images.githubusercontent.com/1578148/58008951-c4af9a00-7b30-11e9-9ef0-584e26b9b2b6.png)

kienstra commented:
Hi @RobStino,
Thanks a lot for testing this.

Good catch. It's definitely not ideal to have that different front-end output when the User (or another pro field) is disabled.

I [see that also](https://github.com/getblocklab/block-lab/pull/306#issuecomment-493802778). It doesn't look to be caused by this PR. 

But I'll look at how we might fix it. If there isn't time before the release today, maybe we could consider that in a different PR.

kienstra commented:
There's a fix for this that's working locally. I'll push it today.

kienstra committed:
If the pro license expires, still render the pro fields on the front-end the same

Rob raised a point that the pro fields
aren't rendered properly on the front-end.
So bootstrap a pro control for the front-end.

kienstra committed:
Abstract license changes into a method

Change the name of set_valid_license()
to set_license_validity().

kienstra commented:
With https://github.com/getblocklab/block-lab/pull/306/commits/1c26b8ebf6ddccd8179614cdaeb68803e56f9bda, the front-end display of a pro field with an expired pro license will still be the same:

<img width="1436" alt="tax-field" src="https://user-images.githubusercontent.com/4063887/58048269-f8c69300-7b0f-11e9-8d3b-c65063e3e5dd.png">

kienstra commented:
As it is right now, there's no change to existing pro fields when a pro license expires:
* You can edit the field in the block editor. For example, you could change a Taxonomy field value from `foo` to `bar`, and it will save just like before.
* It [still displays](https://github.com/getblocklab/block-lab/pull/306#issuecomment-494129721) on the front-end the same.

But you can't change the block in the `/wp-admin` > Block Lab UI:
<img width="1123" alt="taxonmy-here" src="https://user-images.githubusercontent.com/4063887/58003828-44912200-7aa7-11e9-97ec-b5c6cba1e806.png">

lukecarbis commented on php/post-types/class-block-post.php:
```diff
@@ -87,19 +98,25 @@ public function register_controls() {
-				)
-			);
+			foreach ( $this->pro_controls as $pro_control ) {
+				$controls[ $pro_control ] = $this->instantiate_pro_control( $pro_control );
```
Instead of having a (clever) method to `instantiate_pro_control()`, why not just set assign the `pro_controls` to an empty array when they're being defined, and the here do something like:

```
$this->pro_controls = array(
    'post' => new Controls\Post(),
    'taxonomy' => new Controls\Taxonomy(),
    'user' => new Controls\User(),
);
if ( block_lab()->is_pro() ) {
    $controls = array_merge( $controls, $this->pro_controls );
}
```

lukecarbis replied:
Surely this classes `init` method is early enough?

kienstra replied:
Yeah, that's an interesting idea. 

It would make the codebase a little cleaner. And it wouldn't add much overhead. 

But I thought that because most plugin users don't have the pro license, it'd be best to not instantiate the pro controls when they're not going to have access to them. 

`instantiate_pro_control()` is used in a few places to instantiate a pro control just-in-time.

kienstra replied:
Still, your idea would make the codebase cleaner, a pretty strong consideration. My opinion above isn't strongly-held.

lukecarbis replied:
Great thinking. I think the part I don't like is the part where we have to pass the class name as an argument (e.g. `instantiate_control( 'Taxonomy' )`. Feels like a hack to me, especially with the `ucwords`.

What if we:

- Change the method name to `instantiate_control()` since it's not really pro specific anyway
- Even simpler, `get_control()`?
- Check to see that the class _can_ be instantiated before we instantiate it?
- Check to see if the class _is already_ instantiated before we instantiate it?

I'm not sure how all that would fit together – do you think there's some improvements we can pull out from there?

kienstra replied:
Sure, I'm working on this now.

kienstra replied:
@lukecarbis, how does https://github.com/getblocklab/block-lab/commit/addc1afeefcb5c777638bd484b5a1e476faa0242 look? It'll require some regression testing to make sure the URL control still works, as this renames the class from `URL` to `Url`.

lukecarbis commented on php/post-types/class-block-post.php:
```diff
@@ -578,7 +626,15 @@ public function render_fields_meta_box_row( $field, $uid = false ) {
-								data-sync="block-fields-label_<?php echo esc_attr( $uid ); ?>" />
+								data-sync="block-fields-label_<?php echo esc_attr( $uid ); ?>"
+								<?php echo $is_field_disabled ? 'readonly="readonly"' : ''; ?> />
+							<?php if ( $is_field_disabled ) : ?>
```
The indentation here is a little funky…

kienstra replied:
Sure, does [this look better](https://github.com/getblocklab/block-lab/pull/306/commits/11ff4711a7f1e154092900b69c54fc805e65e198#diff-018b57db3c29f17ff59f382617e39f14R631)?

lukecarbis replied:
Yes!

kienstra replied:
Nice, thanks

lukecarbis reviewed (changes requested):
Looks great Ryan!

RobStino commented:
Tested and confirmed that values persist in the editor and the frontend whether Pro is active or not. Thanks Ryan.

kienstra committed:
Use Luke's CSS for the readonly

Also, move the closing /> to the next line,
and move indent one tab to the left.

kienstra commented:
Hi @lukecarbis, 
>Makes sense. Maybe we could add some CSS along the lines of...

Great idea, https://github.com/getblocklab/block-lab/pull/306/commits/11ff4711a7f1e154092900b69c54fc805e65e198 uses your CSS:

<img width="985" alt="input-readonly" src="https://user-images.githubusercontent.com/4063887/58058374-54077e00-7b2e-11e9-879f-a1077ad9b693.png">

kienstra commented:
Hi @lukecarbis,
>Have you tested this out with checkboxes?

Like changing the `<input type="hidden">` [here](https://github.com/getblocklab/block-lab/pull/306#discussion_r285797750) into a checkbox?

lukecarbis commented:
@kienstra No, I mean checking if checkboxes submit the correct value if they are disabled (like selects). But I just realised that we're not loading all the additional options anyway, so it doesn't matter. Please ignore.

kienstra committed:
Change make instantiate_pro_control more general

As Luke suggested, this can be used for non-pro controls.
So use it to instantiate any control,
and first check that the class exists.

kienstra commented on php/post-types/class-block-post.php:
```diff
@@ -81,40 +81,57 @@ public function register_hooks() {
+		$control_names = array(
+			'text',
+			'textarea',
+			'url',
```
To make this work with `ucwords()`, this control class [is renamed](https://github.com/getblocklab/block-lab/pull/306/commits/addc1afeefcb5c777638bd484b5a1e476faa0242#diff-032d8fb82d07a7fc45dc4fd064a36ca8R3) from `Url` to `URL`.

kienstra commented on php/post-types/class-block-post.php:
```diff
@@ -81,40 +81,57 @@ public function register_hooks() {
-			'range'       => new Controls\Range(),
-			'checkbox'    => new Controls\Checkbox(),
-			'radio'       => new Controls\Radio(),
+		$control_names = array(
```
This could have been a property, but I don't think it's needed outside this function. Also, it might add to confusion, as there are `$controls` and `$pro_controls` properties.

kienstra committed:
Fix the DocBlock, as get_control() doesn't only operate on pro controls

Remove references to pro,
as this is now for any control.

kienstra commented:
**Basic Regression Testing**

The Url control looks to still work the same, before and after this PR:

# Before

<img width="1044" alt="before-pr-306" src="https://user-images.githubusercontent.com/4063887/58060968-971b1e80-7b39-11e9-8592-96b9468b598c.png">

# After
<img width="1170" alt="after-pr-306" src="https://user-images.githubusercontent.com/4063887/58060976-a1d5b380-7b39-11e9-97d9-5a7762cdb235.png">

lukecarbis commented on php/post-types/class-block-post.php:
```diff
@@ -81,40 +81,57 @@ public function register_hooks() {
+	 * @return object|null The instantiated pro control.
 	 */
-	public function instantiate_pro_control( $control_name ) {
+	public function get_control( $control_name ) {
```
I like this method a lot better now.

kienstra replied:
Thanks!

lukecarbis reviewed (approved):
Great work!
